### PR TITLE
Refactor location snapping

### LINKED ISSUE #373: Extract location snapping logic and move to core
Location snapping is a core building block for navigation and should be available in Core Navigation.

cc @bsudekum @frederoni

### LINKED ISSUE #406: snapsUserLocationAnnotationToRoute has no effect
`RouteController.snapsUserLocationAnnotationToRoute` purports to affect whether a map view’s user location annotation is snapped to the route. However, Core Navigation doesn’t actually do anything in response to this property; it only affects how RouteMapViewController behaves in the main navigation SDK.

In the Swift example application, CustomViewController sets `snapsUserLocationAnnotationToRoute` to `true` with the expectation that it snaps the user puck to the route, but it doesn’t. Instead, CustomViewController implements its own snapping in `progressDidChange(_:)`. So setting `snapsUserLocationAnnotationToRoute` is a bit misleading.

Additionally, the documentation claims that `snapsUserLocationAnnotationToRoute` is disabled by default, but it’s actually enabled by default.

/cc @bsudekum @ericrwolfe

### LINKED ISSUE #407: Current road name label ignores snapped course
`RouteMapViewController.navigationMapView(_:shouldUpdateTo:)` updates the current road name label after snapping the user’s location to the route line but before snapping the user’s course to the route line. This could cause the label to unexpectedly display the name of a cross street in the event that the course reading is inaccurate.

/cc @bsudekum @frederoni

### PR DESCRIPTION
**Pull request by 1ec5**
This PR moves the user location snapping code from RouteMapViewController in MapboxNavigation to RouteController in CoreNavigation. Conversely, it moves the `snapsUserLocationAnnotationToRoute` property from RouteController in CoreNavigation to RouteMapViewController in MapboxNavigation, where it’s actually used.

RouteController now has public `location` and `snappedLocation` properties, while NavigationViewController has a `snapsUserLocationAnnotationToRoute` property.

Along the way, the current road name label now respects course snapping, and various NavigationMapViewDelegate methods have proper Objective-C names. (Previously, selector pieces ended in dangling prepositions instead of nouns.)

#402 partly refactors the location snapping code as well. I think this PR should land before that one to keep things simple.

To summarize:

* [x] Move user location snapping to Core Navigation (to fix #373)
* [x] Move `snapsUserLocationAnnotationToRoute` property to NavigationViewController (to fix #406)
* [x] Fix current road name label with respect to course snapping (to fix #407)
* [x] Keep labeling the current road even after the user goes off-course
* [x] Fix Objective-C names of NavigationMapViewDelegate methods

/ref #321
/cc @ericrwolfe @frederoni @bsudekum

### CODE REVIEW

File: MapboxNavigation/RouteMapViewController.swift [hunk 1]
```diff
@@ -273,131 +280,100 @@ extension RouteMapViewController: NavigationMapViewDelegate {
...
-                    $0.isMapboxStreets
-            }
-            
-            // Add Mapbox Streets if the map does not already have it
-            if streetsSources.isEmpty {
-                let source = MGLVectorSource(identifier: "mapboxStreetsv7", configurationURL: URL(string: "mapbox://mapbox.mapbox-streets-v7")!)
-                style.addSource(source)
-                streetsSources.append(source)
-            }
+        guard let snappedLocation = routeController.snappedLocation else {
```
Comment (frederoni) [inline MapboxNavigation/RouteMapViewController.swift]: Relying on two location managers, route controller being the main source for driving navigation and MGLMapView asking for `shouldUpdateTo` based on its own location updates feels like it's set up for a race condition. However, #402 should eventually move the data in the other direction and the way CLLocationManager is shared throughout the OS should prevent that but there's no guarantee that the snapped location won't lag 1 location update behind.

Comment (1ec5) [inline MapboxNavigation/RouteMapViewController.swift]: Ah, so should this method be passing `location` into `snappedLocation` then?

Comment (frederoni) [inline MapboxNavigation/RouteMapViewController.swift]: I'd say the other way around, the route controller should drive the map view, but that might be out of scope for this PR?

Comment (1ec5) [inline MapboxNavigation/RouteMapViewController.swift]: Yes, I agree. We should have the route controller drive the map view, and #402 will accomplish that pretty cleanly. In the meantime, do you think what’s happening here is any riskier than what’s in master?

Comment (frederoni) [inline MapboxNavigation/RouteMapViewController.swift]: It should be the same.

File: MapboxCoreNavigation/RouteController.swift [hunk 2]
```diff
@@ -165,6 +159,62 @@ open class RouteController: NSObject {
         locationManager.stopUpdatingLocation()
         locationManager.stopUpdatingHeading()
     }
+    
+    /**
+     The most recently received user location.
+     
+     This is a raw location received from `locationManager`. To obtain an idealized location, use the `snappedLocation` property.
+     */
+    public var location: CLLocation?
```
Comment (bsudekum) [inline MapboxCoreNavigation/RouteController.swift]: Do you think this should be public? If a developer needs the raw user location, shouldn't they just create their own location manager?

Comment (1ec5) [inline MapboxCoreNavigation/RouteController.swift]: snappedLocation has to be public so that MapboxNavigation can use it. I figured if we expose snappedLocation, we might as well expose the raw location too. But we could instead expose snappedLocation as location and pretend there isn't a raw location to begin with.

Comment (bsudekum) [inline MapboxCoreNavigation/RouteController.swift]: ok, that makes sense.

Comment (1ec5) [inline MapboxCoreNavigation/RouteController.swift]: Fixed in 29da08e99177d1f685db285638729d9005485052.

**Comment by 1ec5** [comment]
29da08e99177d1f685db285638729d9005485052 causes the current road name to continue to be labeled even if the user diverges from the route.
